### DISCUSSION TITLE
Reusable workflows and caching

### DISCUSSION THREAD

**Discussion by p3lim** [community/community#26317] [Actions]
<p>I have this envisioned reusable workflow that I’d like to get working, where I need to build binaries on 3 different architectures that will be embedded in my main project. To keep it clean between different workflows I want to re-use a workflow to build these binaries which will be used in 3 other workflows, and use cache to only build when neccessary.</p>
<p>These workflows have not been tested yet, they’re dry-coded to see if it’s possible.</p>
<pre><code class="lang-auto"># .github/workflows/reusable-build-deps.yaml
name: reusable build deps
on:
  workflow_call:

env:
  DEPS_REF: ...

jobs:
  build-x86:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
        with:
          repository: different/project
          ref: ${{ env.DEPS_REF }}

      - uses: actions/cache@v3
        id: cache
        with:
          key: deps-x86-${{ env.DEPS_REF }}
          path: bin/

      - if: ${{ steps.cache.outputs.cache-hit != 'true' }}
        run: |
          # this will output into bin/
          make ...

  build-arm32:
    # same as above except cache key is "deps-arm32-${{ env.DEPS_REF }}" and bin/ contains other files

  build-arm64:
    # same as above except cache key is "deps-arm64-${{ env.DEPS_REF }}" and bin/ contains other files
</code></pre>
<pre><code class="lang-auto"># .github/workflows/build-project.yaml
name: build main project
on:
  push:

jobs:
  build-deps:
    uses: ./.github/workflows/reusable-build-deps.yaml

  build-project:
    - uses: actions/checkout@v3

    - uses: actions/cache@v3
      with:
        key: ???
        path: bin/

    - run: |
        # this will build my main project
        make ...
</code></pre>
<p>Questions, if this is possible:</p>
<ol>
<li>Will the “build-deps” job in my main project workflow call all 3 jobs in the reusable workflow?</li>
<li>Will the “build-deps” jobs be run concurrently, and if so, how can I make sure my “build-project” job waits for it?</li>
<li>How can I get the cached “bin/” combined from all 3 jobs in the reusable workflow into my main workflow?</li>
</ol>

**Comment by airtower-luna** [comment, marked as answer]
<aside class="quote no-group" data-username="p3lim" data-post="1" data-topic="254570">
<div class="title">
<div class="quote-controls"></div>
 p3lim:</div>
<blockquote>
<p>Will the “build-deps” job in my main project workflow call all 3 jobs in the reusable workflow?</p>
</blockquote>
</aside>
<p>Yes.</p>
<aside class="quote no-group" data-username="p3lim" data-post="1" data-topic="254570">
<div class="title">
<div class="quote-controls"></div>
 p3lim:</div>
<blockquote>
<p>Will the “build-deps” jobs be run concurrently, and if so, how can I make sure my “build-project” job waits for it?</p>
</blockquote>
</aside>
<p>Probably. As long as there’s no <code>needs</code> relationship between jobs they’ll be scheduled independently and run concurrently if there are enough runners. To make your “build-project” job wait, add <code>needs: build-deps</code> to it. The documentation <a href="https://docs.github.com/en/actions/using-workflows/reusing-workflows#using-outputs-from-a-reusable-workflow">has an example</a>.</p>
<aside class="quote no-group" data-username="p3lim" data-post="1" data-topic="254570">
<div class="title">
<div class="quote-controls"></div>
 p3lim:</div>
<blockquote>
<p>How can I get the cached “bin/” combined from all 3 jobs in the reusable workflow into my main workflow?</p>
</blockquote>
</aside>
<p>You don’t, caching is the wrong tool here. You need artifacts to reliably transfer files between jobs, see the documentation on <a href="https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts#passing-data-between-jobs-in-a-workflow">passing data between jobs in a workflow</a>. Create an artifact in each dependency job, retrieve all three in the project build. I haven’t tried using artifacts with reusable workflows myself, so I can’t say if there are any pitfalls, but it should work. :slightly_smiling_face:</p>
